### ISSUE TITLE
LotusScript Compilation and Validation

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1234]
While working on #1009, @feather812002 started [looking into options for how to validate the LotusScript used in some of the form options](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1875916697)

>@JoelProminic 
>
>About the compile LotusScript, I think ODP already to do similar work about this :  
>
>> These libraries are brought in with their source stored as [normal text items](https://github.com/OpenNTF/org.openntf.nsfodp/blob/e595156c49cda91f74e344869ee70c616f2d0b39/eclipse/plugins/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L901) named $ScriptLib, but then need to be compiled from there. There's no mechanism for compiling LotusScript in the normal Java API, and IBM's NAPI doesn't have a binding to the [NSFNoteLSCompileExt](http://www-12.lotus.com/ldd/doc/domino_notes/8.5.3/api853ref.nsf/f961929dc6dc14dc852561bf0067d1c4/fa422b61d9afd014482570bb0056c05d?OpenDocument) function involved, so I had to [dip into Java C API bindings](https://github.com/OpenNTF/org.openntf.nsfodp/blob/e595156c49cda91f74e344869ee70c616f2d0b39/eclipse/plugins/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L929), initially via Karsten Lehmann's excellent [Domino JNA](https://github.com/klehmann/domino-jna) and then switched over to [Darwino's NAPI implementation](https://github.com/darwino/domino-napi).
>
>[More Details](https://frostillic.us/blog/posts/2019/7/2/deab5e75ce7c5bdf8525842b004f6c00)
>
> [NSFNoteLSCompile API Document](https://opensource.hcltechsw.com/domino-c-api-docs/reference/Func/NSFNoteLSCompile/)
>
>So , this is my design about this , compile Lotus Script and it's  library in the NSF is possible .
>
>We already have a REST interface from NSFConvertProtal VM, and it can build nsf database with ODP.
>
>So, in short , we just change some code inside of the ODP,let it support REST call and return the result to REST,
>Then we can got the compile result and send it to Moonshine-IDE by REST .

>I to so some research about the NsfODP project how to compile the LotusScript inside.
>Now, I almost understand NSFODP at least support follow type element to compile with Lotus Script.
>-  [LotusScript library](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L824)
>-  [Other element that contain Lotus Script also can be compile but it need some setting](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L650)
>-  [Database that contain the library](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L853)
>
>So just call from the Domino JNA and Darwin's NAPI is possible compile the LotusScript and get error return to user.
>I not sure the Moonshine-IDE support the Java class running env or not support, 
>If it can easy support call a Java Class ,I think we can directly put these code into Moonshine-IDE , then we can let user fast get the return after input any code , 
>But if Moonshine-IDE not support or difficult support Java API call, we maybe have to work on the REST interface on VM.
>
>I already starting try to split this code from NSFODP and put it into standalone Java Jar in today,
>But before I done this , we must should decide we work on which way : direct put the Java Jar to Moonshine ? or Just put the Jar to Vagrant VM, then call it by REST ?

My suggestion is to approach this in incremental steps like this:
1. Create a simple JAR that can compile the code.   If we get Notes Client crashes when running this, we'll probably need to focus on a solution running on a Vagrant instance
2. Create a simple socket server that can receive compilcation requests from Moonshine. This should be able to run from either the local machine (if feasible) or a Vagrant Domino instance.

**Comment by feather812002** [comment]
@JoelProminic  I already staring a the new jar project about the Lotus Script compile, 
So, can I upload it to Git as public ?
Or you create a private rep for this at first ?

**Comment by feather812002** [comment]
@JoelProminic 
I almost done a basic Lotus Script on the NSFConvertProtal VM
But I stuck on a trouble, it only can work success one time , 
When user commit the Lotus Script in second time , it always throw error:



> nsf-converter-portal: Jan 15, 2024 12:50:31 PM com.ibm.commons.log.AbstractLogMgr log
>     nsf-converter-portal: SEVERE: Instance of NSFBase child com.darwino.domino.napi.wrap.NSFNote was not freed (ref count 0) during finalization.



I try to fix this on last Friday and Today , 
But It always report this , I try to re-factory code and add more code release and free the notes:


 ```
@Override
    public void doFree(){
        
        if (this.getHandle() > 0L ) {
            try {
                this.api.NSFNoteClose(this.getHandle());
            }
            catch (DominoException e) {
                throw new RuntimeException(e);
            }finally{
                this.free();
            }
        }
        
        super.doFree();
        
    }
```

No work , it report same error :

 

> nsf-converter-portal: Jan 15, 2024 1:58:40 PM com.ibm.commons.log.AbstractLogMgr log
>     nsf-converter-portal: SEVERE: Instance of NSFBase child com.prominic.lotusscript.compile.app.domino.ProminicNNote was not freed (ref count 0) during finalization.

**Comment by feather812002** [comment]
@JoelProminic 
So, in short ,I not sure this error cause the Domino Java Native API , or it cause from the Darwino Domino JAVA-C API ,
is it possible let metest it on the Domino 13 or Domino 14?

Thanks.

**Comment by feather812002** [comment]
@rat-moonshine Hi,devsena, 
I work with `moonshine.editor.text.TextEditor` currently ,
`languageEditor=new TextEditor(null,false);`

Now, I need setting one line to red color , but keep other line color not change, 
is it have any method can to do this ?
Thanks.

**Comment by rat-moonshine** [comment]
@feather812002 , to colorize some of the text we would need to write custom `ILineParser` instead of using default `PlainTextLineParser`. There is no other way to specify colours.

To specify specific lines in colour, you wants to check the console text-view implementation in Moonshine-IDE.

Here's how the custom `TextEditor` class uses a custom `TextLineRenderer` - https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineGUICore/src/actionScripts/plugin/console/view/ConsoleTextEditor.hx .
(Note: this is in Haxe script, but you can write your own in ActionScript, as well)

This editor also uses a custom `LineParser` (https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineGUICore/src/actionScripts/plugin/console/view/ConsoleLineParser.hx), which is currently works to print lines in [three](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineGUICore/src/actionScripts/plugin/console/view/ConsoleTextEditor.hx#L39) different colours:
```
ConsoleLineParser.CL_ERROR => new TextFormat("_typewriter", 12, 0xff6666)
ConsoleLineParser.CL_WARNING => new TextFormat("_typewriter", 12, 0xFFBF0F)
ConsoleLineParser.CL_SUCCESS => new TextFormat("_typewriter", 12, 0x33cc33)
```
(You can even use this `LineParser` without writing a new)

Then you can adds to your text-editor something in this way:

`myTextEditor.appendtext("Hello world!", ConsoleLineParser.CL_SUCCESS);`

**Comment by feather812002** [comment]
@rat-moonshine 
Hi, rat-moonshine , thanks for your help , 
I think I can understand it how to work at now , I try to reuse your code about the [ConsoleTextEditor.hx](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineGUICore/src/actionScripts/plugin/console/view/ConsoleTextEditor.hx)
But, ugh , it look like I at least should reset the color and other some property for it , then it can work with the Lotus script editor interface.

I not sure  it worth overwrite it just because I only want set one line color ., I will discuss this with Joel more details.
Thanks for you help again.

**Comment by rat-moonshine** [comment]
You are not suggested to overwrite anything into `ConsoleTextEditor.hx`, @feather812002 . However, you can create your own version extending the class, in case you create your specific version.

**Comment by JoelProminic** [comment]
@feather812002, could you summarize how you are currently trying to implement this, and how it would fit into Moonshine?   I was discussing this with @JustinProminic today, and I didn't have a good summary in the issue.

**Comment by JoelProminic** [comment]
I setup the LotusScript compiler through the nsf-converter-portal Vagrant script, and I am now getting confirmation messages when saving a LotusScript Object.  However, I have been trying to trigger a compilation error with the "QueryOpen" object, and I haven't seen an error so far:
- Unclosed String literal
- Only a string literal in the method (triggers error in Designer)
- Remove End Sub

I confirmed that these are errors in Domino Designer.  For example:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/9216a6d5-7d00-4f11-b730-c3f291c5908b)

**Comment by JoelProminic** [comment]
The compiler was failing to start for me in the nsf-converter-portal Vagrant instance:

    ERROR:Not found environment var:DYLD_LIBRARY_PATH ,please check!

I made some updates to the start script to avoid this problem.

**Comment by JoelProminic** [comment]
Type | Valid | Validator | Result
-- | -- | -- | -- 
Formula | Yes | Running | Saves with no prompt
LotusScript | Yes | Running | Triggers a prompt on save
JavaScript | Yes | Running | Saves with no prompt
Formula | No | Running | Shows error in popup
LotusScript | No | Running | Shows error in popup
JavaScript | No | Running | Shows error in popup
Formula | N/A | Not Running | Works, no alert
LotusScript | N/A | Not Running | Works, no alert
JavaScript | N/A | Not Running | Works, no alert

I didn't see any errors on the Console while testing.

I'd like to get rid of the prompt when successfully updating a LotusScript value - this is an unnecessary interruption

Otherwise, I think this is good enough for a merge.

**Comment by JoelProminic** [comment]
I confirmed the valid LotusScript case was updated.   This should be ready for a merge as part of  #1009
